ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,20 +61,51 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
+  public List<KsqlNode> locate(
+      final Struct key,
+      final List<IRoutingFilter> routingFilters
+  ) {
+    final KeyQueryMetadata metadata = kafkaStreams
+        .queryMetadataForKey(stateStoreName, key, keySerializer);
+
+    // Fail fast if Streams not ready. Let client handle it
+    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
+      LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
+                stateStoreName, key);
+      throw new MaterializationException(String.format(
+          "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
+    }
 
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
-      return Optional.empty();
+    final HostInfo activeHost = metadata.getActiveHost();
+    final Set<HostInfo> standByHosts = metadata.getStandbyHosts();
+
+    LOG.info("Before filtering: Active host {} , standby hosts {}", activeHost, standByHosts);
+
+    final List<HostInfo> hosts = new ArrayList<>();
+    hosts.add(activeHost);
+    hosts.addAll(standByHosts);
+
+    // Filter out hosts based on liveness and lag filters.
+    // The list is ordered by routing preference: active node is first, then standby nodes
+    // in order of increasing lag.
+    // If heartbeat is not enabled, all hosts are considered alive.
+    final List<KsqlNode> filteredHosts = new ArrayList<>();
+    for (IRoutingFilter routingFilter : routingFilters) {
```

Comment:
You're adding all nodes that pass at least one filter, and those that pass many will be added multiple times.

I think what you want is to require passing all filters for a host to be added to filteredHosts.